butterfree/hooks/hookable_component.py

```diff
@@ -0,0 +1,154 @@
+            )
+        self.__pre_hook = value
+
+    @property
+    def post_hooks(self) -> List[Hook]:
+        """Function steps to trigger after component main functionality."""
+        return self.__post_hook
+
+    @post_hooks.setter
+    def post_hooks(self, value: List[Hook]) -> None:
+        if not isinstance(value, list):
+            raise ValueError("post_hooks should be a list of Hooks.")
+        if not all(isinstance(item, Hook) for item in value):
+            raise ValueError(
+                "All items on post_hooks list should be an instance of Hook."
+            )
+        self.__post_hook = value
+
+    @property
+    def enable_pre_hooks(self) -> bool:
+        """Property to indicate if the component can define pre_hooks."""
+        return self.__enable_pre_hooks
+
+    @enable_pre_hooks.setter
+    def enable_pre_hooks(self, value: List[Hook]) -> None:
+        if not isinstance(value, bool):
+            raise ValueError("enable_pre_hooks accepts only boolean values.")
+        self.__enable_pre_hooks = value
+
+    @property
+    def enable_post_hooks(self) -> bool:
+        """Property to indicate if the component can define post_hooks."""
+        return self.__enable_post_hooks
+
+    @enable_post_hooks.setter
+    def enable_post_hooks(self, value: List[Hook]) -> None:
+        if not isinstance(value, bool):
+            raise ValueError("enable_post_hooks accepts only boolean values.")
+        self.__enable_post_hooks = value
+
+    def add_pre_hook(self, *hooks: Hook) -> HookableComponent:
+        """Add a pre-hook steps to the component.
+
+        Args:
+            hooks: Hook steps to add to pre_hook list.
+
+        Returns:
+            Component with the Hook inserted in pre_hook list.
+
+        Raises:
+            ValueError: if the component does not accept pre-hooks.
+
+        """
+        if not self.enable_pre_hooks:
+            raise ValueError("This component does not enable adding pre-hooks")
+        self.pre_hooks += list(hooks)
+        return self
+
+    def add_post_hook(self, *hooks: Hook) -> HookableComponent:
+        """Add a post-hook steps to the component.
+
+        Args:
+            hooks: Hook steps to add to post_hook list.
+
+        Returns:
+            Component with the Hook inserted in post_hook list.
+
+        Raises:
+            ValueError: if the component does not accept post-hooks.
+
+        """
+        if not self.enable_post_hooks:
+            raise ValueError("This component does not enable adding post-hooks")
+        self.post_hooks += list(hooks)
+        return self
+
+    def run_pre_hooks(self, dataframe: DataFrame) -> DataFrame:
+        """Run all defined pre-hook steps from a given dataframe.
+
+        Args:
+            dataframe: data to input in the defined pre-hook steps.
+
+        Returns:
+            dataframe after passing for all defined pre-hooks.
+
+        """
+        return reduce(
```

Comment:
I like this implementation, the only problem, from my perspective, is that we wouldn't have access to the intermediate steps, just the final result. Suppose that someone would like to debug the steps in order to check the outputs, ```reduce``` wouldn't allow it. A simpler structure, on the other hand, such as a ```for``` loop could provide this functionality . What do you think?

Reply:
I think that there would be no need if we want to see intermediate steps we should access the hooks we set, and not this method. :thinking: